openwisp_firmware_upgrader/upgraders/openwrt.py

```diff
@@ -536,3 +558,42 @@ def _write_checksum(self, checksum):
         raise ReconnectionFailed(
             "Giving up, device not reachable anymore after upgrade"
         )
+
+    def cancel(self):
+        """Cancel the upgrade operation by revoking Celery task and restarting services."""
+        # Revoke the associated Celery task if it exists
+        self._revoke_celery_task()
+
+        # Restart non-critical services if they were stopped
+        self._restart_services_after_cancellation()
+
+    def _revoke_celery_task(self):
+        """Revoke the associated Celery task if it exists."""
+        try:
+            active_tasks = current_app.control.inspect().active()
+            if not active_tasks:
+                return
+
+            for worker_tasks in active_tasks.values():
+                for task in worker_tasks:
+                    if (
+                        task.get("name")
+                        == "openwisp_firmware_upgrader.tasks.upgrade_firmware"
+                        and task.get("args")
+                        and len(task["args"]) > 0
+                        and str(task["args"][0]) == str(self.upgrade_operation.pk)
+                    ):
+                        current_app.control.revoke(task["id"], terminate=True)
+                        return  # Task found and revoked, exit early
+        except Exception as e:
+            self.log(f"Warning: Could not revoke Celery task: {e}")
+
+    def _restart_services_after_cancellation(self):
```

Comment:
As the restarting is performed in `_check_cancellation`, I think this is redundant.